### PR TITLE
Improved the Type Annotations

### PR DESCRIPTION
**Pull request by michel-steuwer**
I now use VS Code and their Pyright type checking implementation which is much better than PyCharm.

I hope this PR does not change behaviour, but I don't have a good way to test this (without using it in ChocoPy).

### CODE REVIEW

File: src/xdsl/ir.py [hunk 1]
```diff
@@ -232,6 +236,12 @@ class Operation:
     parent: Optional[Block] = field(default=None)
     """The block containing this operation."""
 
+    irdl_operand_defs: List[Tuple[str, OperandDef]] = []
```
Comment (math-fehr) [inline src/xdsl/ir.py:240]: Here, you should write `= field(default_factory=list)`.
This is because this sets a mutable variable as default

**Comment by tobiasgrosser** [comment]
@michel-steuwer, is this sth you plan to merge?

File: src/xdsl/dialects/builtin.py [hunk 2]
```diff
@@ -214,8 +216,8 @@ def get_shape(self) -> List[int]:
     @builder
     def from_type_and_list(
             referenced_type: Attribute,
-            shape: List[Union[int, IntegerAttr]] = None) -> VectorType:
-        if shape is None:
+            shape: List[Union[int, IntegerAttr]] = []) -> VectorType:
```
Comment (wence-) [inline src/xdsl/dialects/builtin.py:219]: antipattern of mutable default argument. For type-checking use:
```
shape: Optional[List[Union[...]]
```
`typing.Optional[a]` is just sugar for `Union[NoneType, a]`

**Comment by tobiasgrosser** [comment]
@michel-steuwer, is this sth you still plan to pick up?

File: src/xdsl/irdl.py [hunk 3]
```diff
@@ -242,7 +245,7 @@ class SingleBlockRegionDef(RegionDef):
 
 
 @dataclass(init=False)
-class AttributeDef:
+class AttributeDef(Attribute):
```
Comment (Dinistro) [inline src/xdsl/irdl.py:248]: This breaks the pytests, due to `Attribute` being a frozen class. Is this required for VSCode, or can this be reverted?

**Comment by math-fehr** [comment]
Should we drop this PR and branch? As far as I know, we integrated those changes in #103?

**Comment by math-fehr** [comment]
Closing this in favor of #103 that is now merged

**Pull request closed without merging**